src/bos/server/controllers/v2/sessiontemplates.py

```diff
@@ -63,12 +63,50 @@ def _sanitize_xnames(st_json):
     Returns:
       Nothing
     """
-    if 'boot_sets' in st_json:
-        for boot_set in st_json['boot_sets']:
-            if 'node_list' in st_json['boot_sets'][boot_set]:
-                clean_nl = [_canonize_xname(node) for node in
-                            st_json['boot_sets'][boot_set]['node_list']]
-                st_json['boot_sets'][boot_set]['node_list'] = clean_nl
+    # There should always be a boot_sets field -- this function
+    # is only called after the template has been verified
+    for boot_set in st_json['boot_sets'].values():
+        if 'node_list' not in boot_set:
+            continue
+        boot_set['node_list'] = [_canonize_xname(node) for node in boot_set['node_list']]
+
+
+def _validate_sanitize_session_template(session_template_id, template_data):
+    """
+    Used when creating or patching session templates
+    """
+    # The boot_sets field is required.
+    if "boot_sets" not in template_data:
+            raise ParsingException("Missing required 'boot_sets' field")
+
+    """All keys in the boot_sets mapping must match the 'name' fields in the
+      boot sets to which they map (if they contain a 'name' field).
+    """
+    for bs_name, bs in template_data["boot_sets"].items():
+        if "name" not in bs:
+            # Set the field here -- this allows the name to be validated
+            # per the schema later
+            bs["name"] = bs_name
+        elif bs["name"] != bs_name:
+            raise ParsingException(f"boot_sets key ({bs_name}) does not match 'name' "
+                                   f"field of corresponding boot set ({bs["name"]})")
+
+    """Convert the JSON request data into a SessionTemplate object.
+       Any exceptions raised here would be generated from the model
+       (i.e. bos.server.models.v2_session_template).
+    """
+    SessionTemplate.from_dict(template_data)
+
+    """
+    We do not bother storing the boot set names inside the boot sets, so delete them.
+    We know every boot set has a name field because we verified that earlier.
+    """
+    for bs in template_data["boot_sets"].values():
```

Comment:
Mitch, I don't think I'm fully understanding boot set names here. Can you give me a quick call to discuss this, please?

A boot set looks like this.
{
  "boot_sets": {
    "compute": {   <<<< Here's the name that you want to get rid of because it's never referenced. However, I think it's instructive for people looking at a boot set to know its name.
      "arch": "X86",
      "etag": "90569c01284bf15176c28e8b5880a0a9",
      "kernel_parameters": "ip=dhcp quiet spire_join_token=${SPIRE_JOIN_TOKEN}",
      "node_roles_groups": [
        "Compute"
      ],
      "path": "s3://boot-images/7d29c7b0-c32b-44e4-9e36-3e1134d0444b/manifest.json",
      "rootfs_provider": "cpss3",
      "rootfs_provider_passthrough": "dvs:api-gw-service-nmn.local:300:hsn0,nmn0:0",
      "type": "s3"
    }
  },
  "cfs": {
    "configuration": "compute-24.3.1-csm-1.5.1-rc.1"
  },
  "enable_cfs": true,
  "name": "compute-24.3.1.x86_64-csm-1.5.1-rc.1",
  "tenant": ""
}

For example, I could have two boot sets for computes with different architectures, x86 and ARM. The differing names compute-x86 and compute-ARM, for example, serve to easily disambiguate the two boot sets.

Reply:
We chatted about this. The name referenced above is not the one being removed by this PR. This PR does nothing to prevent templates with multiple boot sets.